pkg/controller/clusterpool/clusterpool_controller.go

```diff
@@ -415,6 +423,106 @@ func (r *ReconcileClusterPool) Reconcile(ctx context.Context, request reconcile.
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileClusterPool) updateInventory(clp *hivev1.ClusterPool, cds []*hivev1.ClusterDeployment, valid bool, state string, logger log.FieldLogger) {
```

Comment:
Okay, I think I have now grokked this function and its intended purpose.

I propose we refactor to make the logic look more like what we do for CDs+claims. Have a poke through [this chunk](https://github.com/openshift/hive/blob/ea63cac7a85032083993849513e53ffe65b3fe83/pkg/controller/clusterpool/clusterpool_controller.go#L293-L306). What I'm suggesting is:
- Make a new class (struct+methods) in collections.go, e.g. `cdcCollection`, that indexes CDCs in useful ways, including
  - list of available CDCs
  - map keyed by CDC name (easy way to cross-reference from a CD)
  - map keyed by CD (easy way to grab all the "in use" ones)
  - list of CDCs with a CD ref pointing to a nonexistent CD (easy way to discover cases we need to "repair" per [comment](https://github.com/openshift/hive/pull/1672/files#diff-8b381d7d7a9f312bc3c30ac9c0224ec253f3241274a7047b29e87418c9e48c3dR489))
  - ...more if you find you need them
- A loader function, e.g. `getAllCDCsForPool()` that creates and populates :arrow_up: 
- A pair of syncer funcs like `cdcs.SyncClusterDeployments()`/`cds.SyncCDCs()` that basically do what this function's first and second loops do, respectively -- except they do it for *all* CDs and CDCs rather than a chunk at a time.
- Call all those things in the chunk referenced above.